### PR TITLE
minor fix: Enforce sorting

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

File: datafusion/core/src/dataframe/mod.rs [hunk 1]
```diff
@@ -3019,11 +3019,11 @@ mod tests {
...
             [
-                "+-----+-----+----+-------+",
-                "| one | two | c3 | total |",
-                "+-----+-----+----+-------+",
-                "| a   | 3   | 13 | 16    |",
-                "+-----+-----+----+-------+"
+                "+-----+-----+-----+-------+",
+                "| one | two | c3  | total |",
+                "+-----+-----+-----+-------+",
+                "| a   | 3   | -72 | -69   |",
```
Comment (ozankabak) [inline datafusion/core/src/dataframe/mod.rs:3025]: Why did this test change? Let's discuss.

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 2]
```diff
@@ -189,12 +189,12 @@ impl PhysicalOptimizerRule for EnforceSorting {
...
-        let adjusted = sort_pushdown.transform_down(pushdown_sorts)?.data;
-
-        adjusted
+        let adjusted = pushdown_sorts(sort_pushdown)?;
+        let plan = adjusted
             .plan
             .transform_up(|plan| Ok(Transformed::yes(replace_with_partial_sort(plan)?)))
-            .data()
+            .data()?;
+        Ok(plan)
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs]: I think we can do the following instead:
```rust
        pushdown_sorts(sort_pushdown)?
            .plan
            .transform_up(|plan| Ok(Transformed::yes(replace_with_partial_sort(plan)?)))
            .data()
```

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 3]
```diff
@@ -513,8 +513,25 @@ fn remove_corresponding_coalesce_in_sub_plan(
...
+            can_remove = true;
+        } else if let Partitioning::RoundRobinBatch(n_out) = repartition.partitioning() {
+            can_remove =
+                *n_out == repartition.input().output_partitioning().partition_count();
+        }
+        if can_remove {
+            new_req = new_req.children.swap_remove(0)
+        }
+    }
+    Ok(new_req)
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs]:
```suggestion
    let mut new_reqs = requirements.update_plan_from_children()?;
    if let Some(repartition) = new_req.plan.as_any().downcast_ref::<RepartitionExec>() {
        let input_partitioning = repartition.input().output_partitioning();
        // We can remove this repartitioning operator if it is now a no-op:
        let mut can_remove = input_partitioning.eq(repartition.partitioning());
        // We can also remove it if we ended up with an ineffective RR:
        if let Partitioning::RoundRobinBatch(n_out) = repartition.partitioning() {
            can_remove |= *n_out == input_partitioning.partition_count();
        }
        if can_remove {
            new_req = new_req.children.swap_remove(0)
        }
    }
    Ok(new_reqs)
```

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 4]
```diff
@@ -540,8 +557,12 @@ fn remove_corresponding_sort_from_sub_plan(
...
 ) -> Result<PlanWithCorrespondingSort> {
     // A `SortExec` is always at the bottom of the tree.
-    if is_sort(&node.plan) {
-        node = node.children.swap_remove(0);
+    if let Some(sort_exec) = node.plan.as_any().downcast_ref::<SortExec>() {
+        if sort_exec.fetch().is_none() {
+            node = node.children.swap_remove(0);
+        } else {
+            // Do not remove the sort with fetch
+        }
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs]:
```suggestion
    if let Some(sort_exec) = node.plan.as_any().downcast_ref::<SortExec>() {
        // Do not remove sorts with fetch:
        if sort_exec.fetch().is_none() {
            node = node.children.swap_remove(0);
        }
```

File: datafusion/core/src/physical_optimizer/sort_pushdown.rs [hunk 5]
```diff
@@ -30,49 +31,80 @@ use crate::physical_plan::sorts::sort::SortExec;
...
     for (child, requirement) in node.children.iter_mut().zip(reqs) {
-        child.data = requirement;
+        child.data = ParentRequirements {
+            ordering_requirement: requirement,
+            fetch: None,
+        };
+    }
+}
+
+pub(crate) fn call_till_not_early_exit(
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_pushdown.rs]: This seems to be called only once, right below. Let's inline it.

File: datafusion/core/src/physical_optimizer/sort_pushdown.rs [hunk 6]
```diff
@@ -30,49 +31,80 @@ use crate::physical_plan::sorts::sort::SortExec;
...
+        .data
+        .ordering_requirement
+        .as_deref()
+        .unwrap_or(&[]);
     let satisfy_parent = plan
         .equivalence_properties()
         .ordering_satisfy_requirement(parent_reqs);
-
-    if let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() {
+    if plan.as_any().downcast_ref::<SortExec>().is_some() {
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_pushdown.rs]: I think we can use `is_sort` here.

File: datafusion/core/src/physical_optimizer/sort_pushdown.rs [hunk 7]
```diff
@@ -82,12 +114,24 @@ pub(crate) fn pushdown_sorts(
...
+                grand_child.data = ParentRequirements {
+                    ordering_requirement: order,
+                    fetch,
+                };
             }
             // Can push down requirements
-            child.data = None;
-            return Ok(Transformed::yes(child));
+            child.data = ParentRequirements {
+                ordering_requirement: Some(required_ordering.clone()),
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_pushdown.rs]: Do we need this clone?

File: datafusion/core/src/physical_optimizer/sort_pushdown.rs [hunk 8]
```diff
@@ -132,6 +181,35 @@ fn pushdown_requirement_to_children(
...
+            .properties()
+            .eq_properties
+            .requirements_compatible(parent_required, &sort_req)
+        {
+            debug_assert!(!parent_required.is_empty());
+            Ok(Some(vec![Some(parent_required.to_vec())]))
+        } else {
+            Ok(None)
+        }
+    } else if is_limit(plan) {
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_pushdown.rs]: Should this be `is_limit`, or should it push down any operator that has `fetch` and `supports_limits_pushdown` is `true`?

**Review by ozankabak — COMMENTED** [review]
LGTM. I think we should rename the `remove_corresponding_coalesce_in_sub_plan` function because it also removes ineffective repartition operators. Let's update its docstring and find it a new name. Maybe `reoptimize_partitioning`?

File: datafusion/core/src/physical_optimizer/sort_pushdown.rs (hunk 7 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/sort_pushdown.rs]: No, removed it

**Review by ozankabak — APPROVED** [review]
Ready to go

**Comment by ozankabak** [comment]
Merged upstream.

**Pull request closed without merging**